### PR TITLE
42 heatmap filter

### PR DESCRIPTION
**Pull request by vnoelifant**
Summary of changes:

- Normalized data so that the intensity colors made more sense (i.e. Red means more funding is needed for bigger weight, blue means lesser funding needed for lower weight). 
- Also implemented most feedback from PR #41. Would like to add https://github.com/vnoelifant/charity-finder/pull/41#discussion_r1116976879 in the agenda if I can't figure that out. 
- Fourth image shows error where cannot open url from embed HTML code (But it does open on a different tab). 
-  Worked on #42 but maybe would like to display map based on user-selected region, rather than just showing one Region. 

heat_weight_1

heat_weight_2

heat_weight_3

url_not_opening_popup

### CODE REVIEW

File: charity_finder/views.py [hunk 1]
```diff
@@ -11,32 +12,58 @@
...
+        "project_map": project_map,
+    }
+    return render(request, "home.html", context)
+
+
+def get_map():
+
+    # Filtering by remaining funding money by Region
+    # TODO:  Include more regions
+    goal_limit = 100000
```
Comment (bbelderbos) [inline charity_finder/views.py:27]: You can use `_` to make big numbers more readable, e.g. `100_000`, also I usually put constants like this at module level at the top of this file

File: charity_finder/views.py [hunk 2]
```diff
@@ -11,32 +12,58 @@
...
+
+    projects = Project.objects.filter(
+        goal_remaining__gte=goal_limit, region__name="Africa"
```
Comment (bbelderbos) [inline charity_finder/views.py:30]: Africa would also be a constant

File: charity_finder/views.py [hunk 3]
```diff
@@ -11,32 +12,58 @@
...
+    ).values("title", "project_link", "latitude", "longitude", "goal_remaining")
+
+    # For normalizing data for heat map
+    goal_remaining_max = projects.aggregate(Max("goal_remaining"))
+
+    project_map = folium.Map(
+        location=[59.09827437369457, 13.115860356662202], zoom_start=3
```
Comment (bbelderbos) [inline charity_finder/views.py:37]: Wrapping this list in a constant would be more readable, now I am not sure what this lon, lat pair is.

File: charity_finder/views.py [hunk 4]
```diff
@@ -11,32 +12,58 @@
...
             html = """
-                    <b>Project Title:</b>{title} <br>
-                    <a href={url}>Project Link</a>
+                    <b>Project Title:</b>{title}<br>
+                    <a href={url}>Project Link</a><br>
+                    <b>Funding Needed:</b>{goal_remaining}
+                    <b>Funding Weight:</b>{goal_norm}
                     """.format(
-                title=title, url=url
+                title=title, url=url, goal_remaining=goal_remaining, goal_norm=goal_norm
```
Comment (bbelderbos) [inline charity_finder/views.py:66]: you can use an f-string as well

File: charity_finder/views.py [hunk 5]
```diff
@@ -47,15 +74,11 @@ def home(request):
...
-
-    m = m._repr_html_()
+            ).add_to(project_map)
+            HeatMap(lats_longs).add_to(project_map)
 
-    context = {
-        "m": m,
-    }
-    return render(request, "home.html", context)
+    project_map = project_map._repr_html_()
```
Comment (bbelderbos) [inline charity_finder/views.py:80]: Minor stylistic thing ... is that yours or given by the library? If your own I would not start a "public" method with underscore

File: templates/home.html [hunk 6]
```diff
@@ -15,7 +15,7 @@ <h2 class="text-5xl font-bold tracking-tight leading-tight mb-12">
     </div>
     <div class="py-12">
       <h3><b>Underserved Areas</h3></b>
-      {{m | safe}}
+      {{project_map | safe}}
```
Comment (bbelderbos) [inline templates/home.html:18]: much better for variable name

**Review by bbelderbos — APPROVED** [review]

**Comment by bbelderbos** [comment]
Nice work, the site looks amazing in those screenshots!

File: charity_finder/views.py (hunk 1 above)
Comment (bbelderbos) [reply charity_finder/views.py:27]: `GOAL_LIMIT: final = 100_000` -> you can even add a type hint in there stating that we should not override it. You still can but mypy would catch it

File: charity_finder/views.py (hunk 5 above)
Comment (bbelderbos) [reply charity_finder/views.py:80]: ok external, then we cannot change it :)

File: charity_finder/views.py [hunk 7]
```diff
@@ -24,8 +25,12 @@ def get_map():
 
     # Filtering by remaining funding money by Region
     # TODO:  Include more regions
-    GOAL_LIMIT = 100_000
-    REGION_NAME = "Africa"
+    GOAL_LIMIT: Final[
+        int
+    ] = 100_000  # constant marked final, can't assign another value to it
+    REGION_NAME: Final[str] = "Africa"
```
Comment (bbelderbos) [inline charity_finder/views.py:31]: nice, I did a YT video on this recently :)

Comment (vnoelifant) [reply charity_finder/views.py:31]: Awesome! I will check it out. This was so great to learn.

Comment (bbelderbos) [reply charity_finder/views.py:31]: https://www.youtube.com/watch?v=XjQi8v9OPzw > funny you did it better than me by also typing the Final (see comment on video)

**Pull request merged**